core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -816,11 +911,21 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
                                 }
                             }
                         }
+                    } else if (fieldEquals(Fields.PRIMARY_TERMS, currentFieldName)) {
+                        LongArrayList list = new LongArrayList();
+                        while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                            if (token == XContentParser.Token.VALUE_NUMBER) {
+                                list.add(parser.longValue());
+                            } else {
+                                throw new IllegalStateException("found a non-numeric value under [" + Fields.PRIMARY_TERMS.underscore() + "]");
+                            }
+                        }
+                        builder.primaryTerms(list.toArray());
```

Comment:
I did it this way in order not to depend on the json order. We may not know how many shards are there (to allocated the primary terms array) if the settings are parsed after the terms. I felt this is better then an auto growing primary terms array because in other cases you want to have exceptions.